### ISSUE TITLE
Dropdown properties of PropertyGrid are not changing on large windows on ARMv8

### ISSUE THREAD

**Issue by LumaRay** [wxWidgets/Phoenix#1745]
Hello!

I use wxPython on NVIDIA Jetson Nano.
Architecture: ARMv8 x64 (aarch64)
Python 3.6
Ubuntu 18.04LTS
RAM: 4GB+2GB swap
High speed MicroSD as the main drive.
wxPython 4.1.0

I process video frames in real time and have two issues so far:
1. When using PropertyGrid, dropdown properties (enums, booleans), if window size is large, the dropdowns are not changing and change events are not firing, although the dropdowns work without a lag. When I reduce the window size though, the dropdown properties start changing, although with a lag.
2. I have X and Y window position properties that update on window move event. Now although the window moves and redraws smoothly, the move events / PropertyGrid updates are kind of queueing/stacking and firing very slowly with like few seconds delay, independant of the window size. 

I have recorded a screen video of this behavior: https://youtu.be/yZyYKc0tlYo

**Comment by swt2c** [comment]
You're not doing the video processing on the main/GUI thread are you?  If I had to guess, it sounds like the GUI thread is getting starved and isn't able to process events.

**Comment by LumaRay** [comment]
Well, I'm using example code from the Internet... It processes webcam captures and uses timer for that purpose:
```
self.timer = wx.Timer(self)
self.Bind(wx.EVT_TIMER, self.ProcessFrame)
self.timer.Start(50)
```
Actually, I don't really know if it is running on UI or not... Hm....

**Comment by swt2c** [comment]
If `self.ProcessFrame()` is where you are doing the video processing (e.g., work that takes a long time), then yes, that is going to be using the main/UI thread.  That's probably why the UI updates are not happening.  I think you're going to have to do your heavy processing on a separate thread and then periodically communicate the changes to the UI thread when you are ready to display them.

**Comment by LumaRay** [comment]
I've moved image processing to a sepatate thread, from which I still have to query UI elements, like reading propertygrid and setting bmp.
Now PG elements work better, but a little lag still exists as I move window around, X and Y PG values keep changing after I stop moving the window, although for a lesser time.
Strange thing: now the window freezes after few seconds. I can see one processor core busy at 100% at that moment. And I can only resolve this by force stopping the program.
Besides, in this ARM version event still doesn't fire when I change enum dropbox value.

**Comment by RobinD42** [comment]
>  from which I still have to query UI elements

That can still be problematic. There should be no interaction with UI elements from anything other than the UI thread. If you do need to cross the thread barrier, then you'll want to find a way to do it without actually touching the UI elements from the worker thread. For example, using `wx.CallAfter`in the worker thread is an easy way to call a method in the context of the UI thread that does the interaction with the UI elements. The `wx.lib.delayedresult` module provides something a little more powerful. If a worker thread needs to get data from a UI widget then it would be a good idea to pull the data out into some structure that is accessible from either, and have added code that updates the structure when the widget's value changes, or that updates the widget if the value in the data structure changes. Something like PyPubSub is good for things like that. Etc., etc.

**Comment by LumaRay** [comment]
Thanks Robin, I have moved all UI queries to a timer event, then filled a structure there, and use this structure to read current values from image processing threads. This helps a little, though UI is still laggy and sometimes darkens for few seconds, although I've doublechecked that UI is not queried from the threads... Maybe the system limits processing resources for an application, so when I do heavy processing on one of the app's threads, then the UI thread's processing resources become limited as well...

**Comment by RobinD42** [comment]
Yep, that's a general limitation of CPython. Resource-heavy threads can starve other threads unless they release the GIL.
